tests/test_optimizer/test_dist_came.py

```diff
@@ -1,308 +1,564 @@
+    dist_optim = DistributedCAME(tp_param_group, lr=1e-3)
 
     # Setup distributed optimizer
     if zero_size > 1:
-        optim = LowLevelZeroOptimizer(
-            optim,
+        dist_optim = LowLevelZeroOptimizer(
+            dist_optim,
             overlap_communication=True,
             initial_scale=128,
             partition_grad=True,
             dp_process_group=dp_group,
             verbose=True,
         )
-        shard_to_param = optim._param_store.master_to_working_param
-        optim.optim.setup_distributed(tp_group, dp_group, shard_to_param, zero_flag=True)
+        shard_to_param = dist_optim._param_store.master_to_working_param  # {id(): param tensor} but flattened
+        dist_optim.optim.setup_distributed(
+            tensor_parallel_group=tp_group,
+            data_parallel_group=dp_group,
+            shard_to_param=shard_to_param,
+            use_zero=use_zero,
+        )
+    else:
+        shard_to_param = set_master_param_to_shard_param(tp_param_group)
+        dist_optim.setup_distributed(
+            tensor_parallel_group=tp_group,
+            data_parallel_group=dp_group,
+            shard_to_param=shard_to_param,
+            use_zero=use_zero,
+        )
+
+    # ==============================
+    # Correctness Verify
+    # ==============================
+    seed_all(1024)
+    x = torch.randn(WIDTH, HEIGHT, device=local_rank)
+
+    out = base_model(x)
+    out_tp = tp_model(x)
+
+    if zero_size > 1:
+        dist_optim.backward(out_tp.sum())
+        out.sum().backward()
     else:
-        optim.setup_distributed(tp_group, dp_group, None, zero_flag=False)
+        out_tp.sum().backward()
+        out.sum().backward()
+
+    base_optim.step()
+    dist_optim.step()
+
+    base_optim.zero_grad()
+    dist_optim.zero_grad()
+
+    for p, tp_p in zip(base_param_group, tp_param_group):
+        param_is_distributed = is_distributed_tensor(tp_p)
+        if param_is_distributed:
+            shard_spec = get_sharding_spec(tp_p)
+            if len(shard_spec.sharding_sequence) >= 2:
+                # Col Parallel
+                if shard_spec.sharding_sequence[0] == "R":
+                    tp_p = _gather(input_=tp_p, dim=-1, process_group=tp_group)  # gather
+                # ROW Parallel
+                if shard_spec.sharding_sequence[-1] == "R":
+                    tp_p = _gather(input_=tp_p, dim=0, process_group=tp_group)  # gather
+            else:
+                # TP bias
+                tp_p = _gather(input_=tp_p, dim=-1, process_group=tp_group)  # gather
+        else:
+            # No TP bias
+            pass
+        correctness_verify(p.data, tp_p.data, dtype)
+    clear_layout_converter()
+    Randomizer.reset_index()
+    torch.cuda.empty_cache()
+    print(f"Fwd/Bwd Test Pass")
+
+
+@parameterize("dtype", [torch.float16])
+@parameterize("tp_zero_size", [(1, 4)])
+def exam_dist_came_lowlevelzeroplugin(dtype: torch.dtype, tp_zero_size: tuple[int, int]):
```

Comment:
I notice you already check it with zero plugin at line 365. Can we remove this functions?

Reply:
> I notice you already check it with zero plugin at line 365. Can we remove this functions?

Removed in https://github.com/hpcaitech/ColossalAI/pull/5676/commits/e82e2b24a87f8fbae791ebd1cb4e04369bb8b87c;